src/utils.ts

```diff
@@ -0,0 +1,138 @@
+import { PartialFileList } from './PartialFile';
+
+export type StringObject = Record<string, string>;
+
+/**
+ * Type for [[`groupFiles`]] function
+ */
+export type GroupFiles = (
+  flist: PartialFileList | FileList,
+  options: GroupFilesOptions,
+) => GroupedFiles[];
+/**
+ * Group files under same directory path
+ * i.e `[Dir1, Dir2, Dir3...]`  where each Dir is
+ * ```
+ * {
+ *   id,
+ *   extension,
+ *   name,
+ *   captureGroups
+ *   fileA,
+ *   fileB,...
+ *  }
+ *  ```
+ * Files must be in the same directory to be grouped.
+ * @param fileList
+ * @param options - how to build the out object. See `[[GroupFilesOptions]]` for all options.
+ * @return - array of dirs. Some keys are File objects (`fileA`, `fileB`..).
+ */
+
+export const groupFiles: GroupFiles = (
+  fileList,
+  options = { idWithBasename: true, fileKey: 'extension' },
+) => {
+  const { idWithBasename, fileKey, applyToFileKey, captureFromPath } = options;
+
+  type ID = string;
+  let results: Record<ID, Partial<GroupedFiles>> = {};
+
+  for (const file of fileList) {
+    /* name */
+    const dirPath = file.webkitRelativePath.replace(/\/[^/]+$/, '');
+    const filename = applyToFileKey ? applyToFileKey(file.name) : file.name;
+    const basename = filename.replace(/\.[^.]+$/, '');
+    const extension = filename.replace(/.*\./, '');
+
+    /* Group ID */
+    let gid;
+    if (idWithBasename) {
+      gid = `${dirPath}/${basename}`;
+    } else {
+      //just dir path, like for WDF parser.
+      gid = dirPath;
+    }
+
+    /* set the unique grouping id */
+    if (!results[gid]) results[gid] = { gid };
+
+    /* captured information from regex */
+    if (captureFromPath && !results[gid].captureGroups) {
+      const cGroups = dirPath.match(captureFromPath)?.groups;
+      if (cGroups) results[gid].captureGroups = cGroups;
+    }
+
+    /* place file under a particular group `results[gid]` */
+    switch (fileKey) {
+      case 'extension':
+        results[gid][extension] = file;
+        break;
+      case 'basename':
+        results[gid][basename] = file;
+        break;
+      default:
+        results[gid][filename] = file;
+        break;
+    }
+  }
+  return Object.keys(results).map((gid) => results[gid]) as GroupedFiles[];
+};
+
+/**
+ * Interface representing Grouped files as done by [[`groupFiles`]]
+ * there is some metadata and the files (see below).
+ * groupFiles will return an array of these.
+ */
+export interface GroupedFiles {
+  /** group id used to group the files under same path */
+  gid: string;
```

Comment:
I would rename to `key` because is sounds to much like the linux 'gid' that is a number.